### PR TITLE
route config: Allow to configure direct response body size

### LINKED ISSUE #13422: Give the option to increase the maximun size of the HTTP body in responses to be bigger than 4 KB
I know it is for keeping the proxy’s memory footprint from growing too large but it would help to have the option to use a body bigger than 4 KB for the HTTP responses.

### PR DESCRIPTION
**Pull request by dio**
Commit Message: This patch adds `max_direct_response_body_size_bytes` to set the maximum bytes of the direct response body size (in bytes). The config is added as a field in `RouteConfiguration`.

Additional Description: Reviving https://github.com/envoyproxy/envoy/pull/13487 with a slightly different approach (add the config to `RouteConfiguration` instead of directly per direct response config entry).

Risk Level: Low, since the default behavior is preserved.
Testing: Updated to test the newly introduced config.
Docs Changes: Updated.
Release Notes: Added.
Fixes #13422

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

### CODE REVIEW

File: api/envoy/config/route/v3/route.proto
Comment (htuch) [inline api/envoy/config/route/v3/route.proto:117]: Can you be more specific here? I.e. is the concern it's held in configuration memory, or in buffers on the response path, etc.

File: api/envoy/config/route/v3/route.proto
Comment (htuch) [inline api/envoy/config/route/v3/route.proto:112]: You probably want to update `docs/root/intro/arch_overview/http/http_routing.rst` that references 4kb and put in a RST link to this field.

File: source/common/router/config_utility.h
Comment (htuch) [inline source/common/router/config_utility.h:88]: Nit: should go after `api` field.

File: test/common/router/config_impl_test.cc
Comment (htuch) [inline test/common/router/config_impl_test.cc:6172]: What about a positive test? I'm not sure if an integration test makes sense; if we already support one for direct response, I'm thinking the actual size doesn't make any difference there.

**Review by htuch — CHANGES_REQUESTED** [review]
Thanks for the PR @dio! I'll take this review since @mattklein123 is out.
/wait

File: api/envoy/config/route/v3/route.proto
Comment (dio) [reply api/envoy/config/route/v3/route.proto]: Tried to add more description on this. Please let me know if I put it wrongly.

File: api/envoy/config/route/v3/route.proto
Comment (dio) [reply api/envoy/config/route/v3/route.proto:112]: Updated.

File: test/common/router/config_impl_test.cc
Comment (dio) [reply test/common/router/config_impl_test.cc:6195]: Added a case above. And the cases of different data sources combination for setting body content is covered in `RouteMatcherTest.DirectResponse`.

File: source/common/router/config_utility.h
Comment (dio) [reply source/common/router/config_utility.h]: Updated.

File: api/envoy/config/route/v3/route.proto
Comment (htuch) [inline api/envoy/config/route/v3/route.proto:118]: Is this the only reason? My understanding is we also want to avoid flooding watermarked buffers. E.g. if the buffer limits are 128KB and we dump 10MB in a direct response, that will be queued. @alyssawilk for thoughts.

**Review by htuch — COMMENTED** [review]
LGTM modulo comment in `route.proto`. Thanks!
/wait

File: api/envoy/config/route/v3/route.proto
Comment (dio) [reply api/envoy/config/route/v3/route.proto:118]: Oh, this is a gem. Thanks, @htuch! I didn't know that. I will do more digging. And, yes, I'll be very thankful if @alyssawilk can give some thoughts on this.

Comment (alyssawilk) [reply api/envoy/config/route/v3/route.proto:118]: I don't think it'd be queued, I think we'd disable reading from a nonexistent upstream (and risk OOMs if there were lots of direct responses)
I think it'd be fine to allow larger responses (though I'd be inclined to add a warning note in the API about it not being subject to memory guards so use with care) but suggest an integration test to make sure it all works as I think it should :-)

Comment (dio) [reply api/envoy/config/route/v3/route.proto:118]: I have added "standard" cases to the direct response integration test. If you think we need to cover more cases, please let me know. re: Buffer limit, I have a test case that sends 1000 * 4K so I think it's enough.

File: api/envoy/config/route/v3/route.proto
Comment (htuch) [inline api/envoy/config/route/v3/route.proto:119]: I might phrase this as "data plane buffering controls" rather than "memory guards".

**Review by htuch — COMMENTED** [review]
LGTM, but defer to @alyssawilk for a pass/merge.

File: api/envoy/config/route/v3/route.proto
Comment (dio) [reply api/envoy/config/route/v3/route.proto]: Updated.

File: api/envoy/config/route/v3/route.proto
Comment (alyssawilk) [inline api/envoy/config/route/v3/route.proto:119]: s/being//

File: docs/root/version_history/current.rst
Comment (alyssawilk) [inline docs/root/version_history/current.rst:56]: s/is/remains/ so folks can easily tell there's no change?

File: test/integration/direct_response_integration_test.cc
Comment (alyssawilk) [inline test/integration/direct_response_integration_test.cc:61]: mind also setting 
config_helper_.setBufferLimits(1024, 1024);
for this one?

**Review by alyssawilk — COMMENTED** [review]
Looks good overall - I'm sure this will be really useful to a lot of users - thanks for adding it!

File: test/integration/direct_response_integration_test.cc
Comment (dio) [reply test/integration/direct_response_integration_test.cc:62]: Updated.

File: api/envoy/config/route/v3/route.proto
Comment (dio) [reply api/envoy/config/route/v3/route.proto]: Updated. Thanks for reviewing!

File: docs/root/version_history/current.rst
Comment (dio) [reply docs/root/version_history/current.rst]: 👍🏽

**Review by alyssawilk — DISMISSED** [review]

File: docs/root/version_history/current.rst
Comment (alyssawilk) [inline docs/root/version_history/current.rst:65]: is this a bad merge?

Comment (dio) [reply docs/root/version_history/current.rst]: Yes, sorry. This is updated.

**Review by alyssawilk — APPROVED** [review]

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
